[12.0][IMP] cooperator: backport changes from migration to 14.0

huguesdk opened the pull request:
also includes changes from #370.

## changes

*   fix mail templates: mostly whitespace cleanup, but also fix firstname reference for companies (instead of displaying “false”).
*   remove subscription request `subscription` state (use `new` instead) (was used when creating a subscription request from a partner).
*   fix partner matching by email or company register number when creating a subscription request.
*   fix subscription request name for companies: use company name (like previously) instead of representative name. the name is displayed in the title of the form, and can be used to search.
*   make country and language fields mandatory on the web form.
*   fix name of reports and attachments.
*   some code refactoring and cleanup.
*   improve `fr` and `fr_be` translations.

edit : fix https://github.com/coopiteasy/vertical-cooperative/issues/288

victor-champonnois commented on cooperator/models/partner.py:
```diff
@@ -85,6 +85,8 @@ def _compute_effective_date(self):
         for partner in self:
             if partner.share_ids:
                 partner.effective_date = partner.share_ids[0].effective_date
+            else:
+                partner.effective_date = False
```
Why this change ? because effective_date is already False by default ?

huguesdk replied:
because [compute functions must always set a value](https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-13.0#tasks-to-do-in-the-migration) (in 13.0, but like this we keep differences between 12.0 and 14.0 to a minimum).

victor-champonnois commented on cooperator/models/partner.py:
```diff
@@ -122,12 +124,12 @@ def _compute_share_info(self):
     member = fields.Boolean(
         string="Effective cooperator",
-        help="Check this box if this cooperator" " is an effective member.",
+        help="Check this box if this cooperator is an effective member.",
```
Effective member is the same thing as effective cooperator ? This is misleading no ?

huguesdk replied:
indeed! i actually didn’t read the string, just removed the quotes. we should settle on one, let’s ask the analysts.

victor-champonnois replied:
tagging @polchampion on this

polchampion replied:
@victor-champonnois "effective cooperator" sounds more adequate.

victor-champonnois replied:
@huguesdk do you want to fix this here or to create an issue ?

huguesdk replied:
the string can be fixed here (and then also in #373), but if we want to change the field name, then we need an issue.

victor-champonnois commented on cooperator/models/partner.py:
```diff
@@ -139,7 +141,7 @@ def _compute_share_info(self):
     old_member = fields.Boolean(
         string="Old cooperator",
-        help="Check this box if this cooperator is" " no more an effective member.",
+        help="Check this box if this cooperator is no more an effective member.",
```
We could consider renaming this variable as "former_member", what do you think ? (this would not be done here, but we can create an issue for future komunigi work)

huguesdk replied:
would be better english, indeed. is there a document where all renaming suggestions are listed together? that  would be the place for this, as it would be part of a separate change.

victor-champonnois replied:
Don't know about such a list. Maybe @robinkeunen knows where to keep track of those ?

victor-champonnois replied:
issue opened here https://github.com/coopiteasy/vertical-cooperative/issues/386

robinkeunen replied:
No existing list of possible renamings no. Opening issues is the way to go.

Also, better English yet would be :

> Check this box if this cooperator is not an effective member anymore.

victor-champonnois commented on cooperator/models/partner.py:
```diff
@@ -227,6 +229,11 @@ def get_representative(self):
         return self.child_ids.filtered("representative")
 
     def get_cooperator_from_email(self, email):
+        if not email:
+            return self.browse()
+        email = email.strip()
+        if not email:
+            return self.browse()
```
Why the duplication ? Maybe a comment would be welcome ?

huguesdk replied:
good point. this could also be written:
```python
if email:
    email = email.strip()
if not email:
    return self.browse()
```
or comments added to the current version:
```python
# email could be falsy
if not email:
    return self.browse()
email = email.strip()
# or could be only made of whitespace. this test must be done twice
# because None or False have no .strip() method.
if not email:
    return self.browse()
```
which version do you prefer?

victor-champonnois replied:
The second one is more helpful to the reader than the first one, and could prevent a future error.

huguesdk replied:
what about this?
```python
if email:
    email = email.strip()
# email could be falsy or be only made of whitespace.
if not email:
    return self.browse()
```

victor-champonnois replied:
yes, even better :)

victor-champonnois replied:
actually I still prefer the second one because it is easier to understand, but maybe that's just me ^^

huguesdk replied:
@robinkeunen?

robinkeunen replied:
I like the last version, easier on the eye, less ways out of the function.

carmenbianca replied:
The version with the fewest returns I like best.

victor-champonnois commented on cooperator/models/subscription_request.py:
```diff
@@ -78,54 +80,60 @@ def _check_share_available_to_user(self):
+            ]
+            # we don't use partner.coop_candidate because we want to also
+            # handle draft and waiting requests.
+            if partner.member or self.search(pending_requests_domain):
+                vals["type"] = "increase"
```
I have the impression that self.search(pending_requests_domain) could return something even if the partner is not member.
And if the partner exists but is not member, the type is "new", not "increase" no ?

huguesdk replied:
the logic i used is that if a partner has already made a subscription request (even not validated yet, but not canceled), subsequent subscription requests should be of type `increase`. only the first subscription request would be of type `new`. does that sound logical to you?

victor-champonnois replied:
OK, It is a good option. It is not ideal because there can be cases where the first request will be cancelled later, and the only request will have type "increase". But I guess this case is not very frequent and too costly to account for...?

polchampion replied:
Reading your comments, it looks like the type should ideally be attributed later, when the request is validated (not at creation), at which point it makes more sense to check whether the subscription is new or increase. Just saying, but probably ok to leave it as huguesdk proposes.

huguesdk replied:
very good remarks. the type on the request itself does not make much sense until it is validated. what is this type used for, anyway?

polchampion replied:
I'm not sure.

victor-champonnois commented on cooperator/models/subscription_request.py:
```diff
@@ -143,12 +151,17 @@ def check_iban(self, iban):
-            sub_request.name = " ".join(
-                part for part in (sub_request.firstname, sub_request.lastname) if part
-            )
+            if sub_request.is_company:
+                sub_request.name = self.company_name
```
Wouldn't this overwrite the representative's name ?

huguesdk replied:
yes, but the name of a subscription request for a company is supposed to be the name of the company (as it was in 9 and in 12 before the changes to the `name` field). this allows to search by the company name. `firstname` and `lastname` are still set to the name of the representative.

victor-champonnois replied:
OK :+1:

victor-champonnois reviewed (approved):
LGTM, modulo some questions

huguesdk commented:
@cathLemb or @polchampion: could you please validate this functionally? it is deployed on the test servers (on all dbs, and should work on a new db also, of course). all changes are listed in the description.

cathLemb commented:
Use emc12-demo_test, emc12-demo-testFR and emc12-demo-testCH (new db i've created and installed the modules)
I've done tests on the belgian db (except for tax shelter). Everything seems ok. 
take share for a company/a person - from back end or front end
selling a person share
transfering a person share

- fix mail templates : mostly whitespace cleanup, but also fix firstname reference for companies (instead of displaying “false”).
False is indeed replaced, but the request to pay email is not harmonized with the rest of the template (concerning the whitespaces). I can live with it but I think it's better to harmonize. 

- remove subscription request subscription state (use new instead) (was used when creating a subscription request from a partner).
ok

- make country and language fields mandatory on the web form.
Seems ok but I don't see any difference with what is in prod?

- fix name of reports and attachments.
reports works 

- some code refactoring and cleanup.
difficult to test but everything seems fine. 

- improve fr and fr_be translations.
ok

- fix partner matching by email or company register number when creating a subscription request.
not sure what was the problem and what do I have to test? 

- fix subscription request name for companies : use company name (like previously) instead of representative name. the name is displayed in the title of the form, and can be used to search.
not sure what this is about?

huguesdk committed:
[IMP] backport changes from migration to 14.0

* improve email and report templates: correct some strings, clean up
  some whitespace, and fix references to first name for companies
  (instead of displaying "false").
* remove subscription request "subscription" state (use "new" instead)
  (was used when creating a subscription request from a partner).
* refactor subscription request .create() method to handle both people
  and companies.
* fix partner matching by email or company register number when creating
  a subscription request.
* fix subscription request name for companies: use company name (like
  previously) instead of representative name. the name is displayed in
  the title of the form, and can be used to search.
* add several subscription request tests.
* make country and language fields mandatory on the web form.
* fix name of reports and attachments.
* some code refactoring and cleanup.
* improve fr and fr_be translations.

co-authored-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
co-authored-by: Victor Champonnois <victor@coopiteasy.be>

huguesdk commented:
/ocabot merge major
